alf/drivers/off_policy_driver.py

```diff
@@ -210,155 +145,8 @@ def train(self,
-            exp = tf.nest.map_structure(lambda ta: ta.read(counter),
-                                        experience_ta)
-            collect_action_distribution_param = exp.action_distribution
-            collect_action_distribution = nested_distributions_from_specs(
-                self._action_distribution_spec,
-                collect_action_distribution_param)
-            exp = exp._replace(action_distribution=collect_action_distribution)
-
-            policy_state = common.reset_state_if_necessary(
-                policy_state, initial_train_state,
-                tf.equal(exp.step_type, StepType.FIRST))
-
-            policy_step = common.algorithm_step(self._algorithm.train_step,
-                                                exp, policy_state)
-
-            action_dist_param = common.get_distribution_params(
-                policy_step.action)
-
-            training_info = make_training_info(
-                action=exp.action,
-                action_distribution=action_dist_param,
-                reward=exp.reward,
-                discount=exp.discount,
-                step_type=exp.step_type,
-                info=policy_step.info,
-                collect_info=exp.info,
-                collect_action_distribution=collect_action_distribution_param)
-
-            training_info_ta = tf.nest.map_structure(
-                lambda ta, x: ta.write(counter, x), training_info_ta,
-                training_info)
-
-            counter += 1
-
-            return [counter, policy_step.state, training_info_ta]
-
-        with tf.GradientTape(
-                persistent=True, watch_accessed_variables=False) as tape:
-            tape.watch(self._trainable_variables)
-            [_, _, training_info_ta] = tf.while_loop(
-                cond=lambda counter, *_: tf.less(counter, num_steps),
-                body=_train_loop_body,
-                loop_vars=[counter, first_train_state, training_info_ta],
-                back_prop=True,
-                name="train_loop")
-            training_info = tf.nest.map_structure(lambda ta: ta.stack(),
-                                                  training_info_ta)
-            action_distribution = nested_distributions_from_specs(
-                self._action_distribution_spec,
-                training_info.action_distribution)
-            collect_action_distribution = nested_distributions_from_specs(
-                self._action_distribution_spec,
-                training_info.collect_action_distribution)
-            training_info = training_info._replace(
-                action_distribution=action_distribution,
-                collect_action_distribution=collect_action_distribution)
-
-        loss_info, grads_and_vars = self._algorithm.train_complete(
-            tape=tape, training_info=training_info, weight=weight)
-
-        del tape
-
-        return training_info, loss_info, grads_and_vars
+        return self._algorithm.train(
```

Comment:
off_policy_driver.train() can be removed. Call off_policy_algorithm.train instead